_posts/2024-09-12-opensearch-project-roadmap-2024-2025.md

```diff
@@ -0,0 +1,187 @@
+
+[**Optimized privilege evaluation**](https://github.com/opensearch-project/security/issues/3870): Performance is a key focus for OpenSearch. We’ve identified areas within the Security plugin that can yield significant performance improvements, especially for clusters with numerous indexes or roles mapped to users.
+
+[**API tokens**](https://github.com/opensearch-project/security/issues/4009): API tokens introduce a new way to interact with OpenSearch clusters by associating permissions directly with a token. Cluster administrators will have full visibility and control over the issued tokens and their usage.
+
+[**Ease of use**](https://github.com/opensearch-project/security-dashboards-plugin/issues/2070): We aim to simplify security setup for cluster administrators. Many useful security features remain underused because they are not exposed through OpenSearch Dashboards. To address this, we will add security dashboard pages, where administrators can configure rate limiters to protect clusters from unauthenticated actors.
+
+Looking ahead, security primitives like [authorization could be extracted and made pluggable](https://github.com/opensearch-project/security/issues/4702), allowing integration with newer open standards for policy evaluation, such as Open Policy Agent (OPA) or Cedar.
+
+
+### Roadmap Theme 8: Modular architecture
+
+OpenSearch is working towards well-supported modularity in order to enable **rapid development of properly encapsulated features and flexible deployment architectures** for cloud-native use cases. Historically, OpenSearch has been deployed and operated as a cluster model, in which all functions (such as replication and durability) were implemented within the cluster. While the project has grown organically, offering many extension points through plugins, it still relies on a monolithic server module at the core, with tight coupling across the architecture. As the project grows within a globally distributed community, this monolithic architecture will become an unsustainable bottleneck. Innovations such as the next-generation query engine are not possible with tightly coupled components. Additionally, the Java Security Manager is pending deprecation and removal from the Java runtime, and the recommended replacement technique ([shallow sandboxing](https://inside.java/2021/04/23/security-and-sandboxing-post-securitymanager/)) relies on using [newer language features that require properly modularized code](https://github.com/opensearch-project/OpenSearch/issues/1588). The overall goal of the [modularity effort](https://github.com/opensearch-project/OpenSearch/issues/5910) is to allow the same core OpenSearch code to run across all variants (for example, on-prem clusters and large managed serverless offerings) while providing strong encapsulation of cluster functions. This will facilitate more independent development and innovation across the project.
```

Comment:
```suggestion
OpenSearch is working toward well-supported modularity in order to enable **rapid development of properly encapsulated features and flexible deployment architectures** for cloud-native use cases. Historically, OpenSearch has been deployed and operated as a cluster model, in which all functions (such as replication and durability) were implemented within the cluster. While the project has grown organically, offering many extension points through plugins, it still relies on a monolithic server module at its core, with tight coupling across the architecture. As the project grows within a globally distributed community, this monolithic architecture will become an unsustainable bottleneck. Innovations such as the next-generation query engine are not possible with tightly coupled components. Additionally, the Java Security Manager is pending deprecation and removal from the Java runtime, and the recommended replacement technique, ([shallow sandboxing](https://inside.java/2021/04/23/security-and-sandboxing-post-securitymanager/)), relies on using [newer language features that require properly modularized code](https://github.com/opensearch-project/OpenSearch/issues/1588). The overall goal of the [modularity effort](https://github.com/opensearch-project/OpenSearch/issues/5910) is to allow the same core OpenSearch code to run across all variants (for example, on-premises clusters and large managed serverless offerings) while providing strong encapsulation of cluster functions. This will facilitate more independent development and innovation across the project.
```